cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -602,7 +602,9 @@ Future<std::vector<ExecBatch>> DeclarationToExecBatchesAsync(Declaration declara
       .Then([collected_fut, exec_plan]() -> Result<std::vector<ExecBatch>> {
         ARROW_ASSIGN_OR_RAISE(auto collected, collected_fut.result());
         return ::arrow::internal::MapVector(
-            [](std::optional<ExecBatch> batch) { return std::move(*batch); },
+            [](std::optional<ExecBatch> batch) {
+              return ARROW_PREDICT_TRUE(batch) ? std::move(*batch) : ExecBatch();
```

Comment:
This is fine with me, but could also use https://en.cppreference.com/w/cpp/utility/optional/value_or for concision.

Reply:
> Does value_or use something like ARROW_PREDICT_TRUE underneath?

Probably not, but that probably doesn't matter either.